### ISSUE TITLE
C# 9 Records in the API Template

### ISSUE THREAD

**Issue by RehanSaeed** [Dotnet-Boxed/Templates#877]
# Boxed.Mapping

I recently added support for C# 9 records to the [Boxed.Mapping](https://github.com/Dotnet-Boxed/Framework/pull/238) library, so we can now map to and from record types.

# Swagger Comments

We could convert our view models to records. However, Swagger still needs documentation and there are two ways to add Swagger documentation to a record.

## C# 9 Positional Records & XML Comments

We can't use positional records because there is no way to provide comments for the properties. There is a Roslyn bug open for this at https://github.com/dotnet/roslyn/issues/44571. This would be the ideal solution.

## C# 9 Records & XML Comments

We can create `init` properties instead and comment them as usual.

```cs
/// <summary>A product.</summary>
public record Product
{
    /// <summary>Gets or sets the unique identifier.</summary>
    public int Id { get; init; }
}
```

## C# 9 Records & Swagger Attributes

I tried with attributes instead but this currently doesn't work, see https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1920.

```cs
public record Product([SwaggerSchema("The product identifier")] int Id);
```

**Comment by VictorioBerra** [comment]
Because of the Swagger support, maybe we should wait a bit before implementing this? Also, viewmodels get mutated in controllers all the time, especially during postbacks. Do we really want to make them immutable by default? This kind of goes along with nullable reference types. Its a pretty big paradigm shift in the way code is written. Just because we _can_ map them now, doesn't mean we have to bring that into the default template.

**Comment by RehanSaeed** [comment]
Yes I kind of agree. The PATCH and PUT commands do mutate as you say with the `IMapper<SaveCar, Models.Car>` being the main problem to overcome. There are two ways this can be done:

1. Only move view models to records and leave the Models as mutable classes.
2. Change everything to records and create a new type of mapper. This is the current immutable mapper I added last week:

```cs
public interface IImmutableMapper<TSource, TDestination>
{
    TDestination Map(TSource source);
}
```

We could also add one that accepts an existing destination object:


```cs
public interface IImmutableWithMapper<TSource, TDestination>
{
    TDestination Map(TSource source, TDestination existing);
}

public class Example : IImmutableWithMapper<Source, Destination>
{
    Destination Map(Source source, Destination existing)
    {
        return existing with {
            Foo = source.Foo,
        };
    }
}
```

This would mean that some immutable types would need to implement both `IImmutableMapper<TSource, TDestination>` and `IImmutableWithMapper<TSource, TDestination>` but not sure if this is a good idea or not.

**Comment by VictorioBerra** [comment]
Could we hide it all behind a template flag? `--immutable-viewmodels | -ivms` if not, its easy for consumers of the template to replace this as needed since this is a pretty specific chunk of example business logic.

I will say its a good platform to show off your mapping utilities 😉

**Comment by RehanSaeed** [comment]
I'd rather not make it an option on this feature. We either enable it or not. Just not sure if I should.

**Comment by RehanSaeed** [comment]
Here is my draft PR: https://github.com/Dotnet-Boxed/Templates/pull/883.

**Comment by VictorioBerra** [comment]
Thats a pretty straightforward change. I guess thats a lot cleaner and simpler than I thought it would be. Either way, its easy for someone to change. I wish we could do a poll or something to get feedback.

**Comment by RehanSaeed** [comment]
That doesn't include the changes to the PUT and PATCH commands. That will be nastier and will require changes to the mapper library I explained above. Not sure it's worth it.
